File: src/main/java/me/zeroeightsix/kami/module/modules/movement/Velocity.java
package me.zeroeightsix.kami.module.modules.movement;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.KamiEvent;
import me.zeroeightsix.kami.event.events.EntityEvent;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.network.play.server.SPacketEntityVelocity;
import net.minecraft.network.play.server.SPacketExplosion;

/**
 * Created by 086 on 16/11/2017.
 * @see me.zeroeightsix.kami.mixin.client.MixinBlockLiquid
 */
@Module.Info(
        name = "Velocity",
        description = "Modify knockback impact",
        category = Module.Category.MOVEMENT
)
public class Velocity extends Module {

    private Setting<Boolean> noPush = register(Settings.b("NoPush", true));
    private Setting<Float> horizontal = register(Settings.f("Horizontal", 0));
    private Setting<Float> vertical = register(Settings.f("Vertical", 0));

    @EventHandler
    private Listener<PacketEvent.Receive> packetEventListener = new Listener<>(event -> {
        if (event.getEra() == KamiEvent.Era.PRE) {
            if (event.getPacket() instanceof SPacketEntityVelocity) {
                SPacketEntityVelocity velocity = (SPacketEntityVelocity) event.getPacket();
                if (velocity.getEntityID() == mc.player.entityId) {
                    if (horizontal.getValue() == 0 && vertical.getValue() == 0) event.cancel();
                    velocity.motionX *= horizontal.getValue();
                    velocity.motionY *= vertical.getValue();
                    velocity.motionZ *= horizontal.getValue();
                }
            } else if (event.getPacket() instanceof SPacketExplosion) {
                if (horizontal.getValue() == 0 && vertical.getValue() == 0) event.cancel();
                SPacketExplosion velocity = (SPacketExplosion) event.getPacket();
                velocity.motionX *= horizontal.getValue();
                velocity.motionY *= vertical.getValue();
                velocity.motionZ *= horizontal.getValue();
            }
        }
    });

    @EventHandler
    private Listener<EntityEvent.EntityCollision> entityCollisionListener = new Listener<>(event -> {
        if (event.getEntity() == mc.player) {
            if ((horizontal.getValue() == 0 && vertical.getValue() == 0) || noPush.getValue()) {
                event.cancel();
                return;
            }
            event.setX(-event.getX() * horizontal.getValue());
            event.setY(0);
            event.setZ(-event.getZ() * horizontal.getValue());
        }
    });

}

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/Velocity.kt
package me.zeroeightsix.kami.module.modules.movement

import me.zeroeightsix.kami.event.events.EntityCollisionEvent
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.mixin.client.world.MixinBlockLiquid
import me.zeroeightsix.kami.mixin.extension.packetMotionX
import me.zeroeightsix.kami.mixin.extension.packetMotionY
import me.zeroeightsix.kami.mixin.extension.packetMotionZ
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import net.minecraft.network.play.server.SPacketEntityVelocity
import net.minecraft.network.play.server.SPacketExplosion
import org.kamiblue.event.listener.listener

/**
 * @see MixinBlockLiquid.modifyAcceleration
 */
object Velocity : Module(
    name = "Velocity",
    description = "Modify knock back impact",
    category = Category.MOVEMENT
) {
    private val noPush = setting("NoPush", true)
    private val horizontal = setting("Horizontal", 0f, -5f..5f, 0.05f)
    private val vertical = setting("Vertical", 0f, -5f..5f, 0.05f)

    init {
        listener<PacketEvent.Receive> {
            if (mc.player == null) return@listener
            if (it.packet is SPacketEntityVelocity) {
                with(it.packet) {
                    if (entityID != mc.player.entityId) return@listener
                    if (isZero) {
                        it.cancel()
                    } else {
                        packetMotionX = (packetMotionX * horizontal.value).toInt()
                        packetMotionY = (packetMotionY * vertical.value).toInt()
                        packetMotionZ = (packetMotionZ * horizontal.value).toInt()
                    }
                }
            } else if (it.packet is SPacketExplosion) {
                with(it.packet) {
                    if (isZero) {
                        it.cancel()
                    } else {
                        packetMotionX *= horizontal.value
                        packetMotionY *= vertical.value
                        packetMotionZ *= horizontal.value
                    }
                }
            }
        }

        listener<EntityCollisionEvent> {
            if (it.entity != mc.player) return@listener
            if (noPush.value || isZero) {
                it.cancel()
            } else {
                it.x = it.x * horizontal.value
                it.y = it.y * vertical.value
                it.z = it.z * horizontal.value
            }
        }
    }

    private val isZero get() = horizontal.value == 0f && vertical.value == 0f
}